File: src/packages/control-ui/src/api/mesh_wireguard_coordinator.ts
// Control UI TypeScript: WireGuard Mesh Coordinator API

export type MeshConnectionMode = 'direct' | 'derp_relay';

export interface MeshPeer {
  peerId: string;
  publicKeyHex: string;
  virtualIp: string;
  allowedIps: string[];
  endpoints: string[];
  derpRegionId: number;
  lastHandshakeMs: number;
  isExitNode: boolean;
}

export class MeshWireguardCoordinator {
  public localPeerId: string;
  public localVirtualIp: string;
  public peers: Map<string, MeshPeer> = new Map();
  public derpRelays: Map<number, string> = new Map();

  constructor(localPeerId: string, localVirtualIp: string) {
    this.localPeerId = localPeerId;
    this.localVirtualIp = localVirtualIp;
  }

  public registerDerpRelay(regionId: number, hostname: string): void {
    this.derpRelays.set(regionId, hostname);
  }

  public registerPeer(peer: MeshPeer): void {
    this.peers.set(peer.peerId, peer);
  }

  public selectBestEndpoint(
    peerId: string,
    nowMs: number,
  ): { mode: MeshConnectionMode; endpoint: string } | null {
    const peer = this.peers.get(peerId);
    if (!peer) return null;

    const isRecent = nowMs - peer.lastHandshakeMs < 180_000;
    if (isRecent && peer.endpoints.length > 0) {
      return { mode: 'direct', endpoint: peer.endpoints[0]! };
    }

    if (this.derpRelays.has(peer.derpRegionId)) {
      return {
        mode: 'derp_relay',
        endpoint: this.derpRelays.get(peer.derpRegionId)!,
      };
    }

    if (peer.endpoints.length > 0) {
      return { mode: 'direct', endpoint: peer.endpoints[0]! };
    }

    return null;
  }

  public lookupRoute(destIp: string): string | null {
    for (const [id, peer] of this.peers.entries()) {
      if (peer.virtualIp === destIp) return id;
    }

    for (const [id, peer] of this.peers.entries()) {
      for (const cidr of peer.allowedIps) {
        if (this.isIpInCidr(destIp, cidr)) return id;
      }
    }

    return null;
  }

  public generateWireguardConfig(peerId: string): string {
    const peer = this.peers.get(peerId);
    if (!peer) throw new Error(`Peer not found: ${peerId}`);

    const allowed =
      peer.allowedIps.length > 0 ? peer.allowedIps.join(', ') : `${peer.virtualIp}/32`;
    const lines = ['[Peer]', `PublicKey = ${peer.publicKeyHex}`, `AllowedIPs = ${allowed}`];
    if (peer.endpoints.length > 0) {
      lines.push(`Endpoint = ${peer.endpoints[0]}`);
      lines.push('PersistentKeepalive = 25');
    }
    return lines.join('\n');
  }

  private isIpInCidr(ip: string, cidr: string): boolean {
    const [netStr, prefixStr] = cidr.split('/');
    if (!netStr || !prefixStr) return false;
    const prefix = parseInt(prefixStr, 10);
    const ipNum = this.ipToNumber(ip);
    const netNum = this.ipToNumber(netStr);
    const mask = prefix === 0 ? 0 : (~0 << (32 - prefix)) >>> 0;
    return (ipNum & mask) === (netNum & mask);
  }

  private ipToNumber(ip: string): number {
    return ip.split('.').reduce((acc, octet) => ((acc << 8) + parseInt(octet, 10)) >>> 0, 0);
  }
}

File: labs/mobile/donor-archive/tunnel/MeshWireguardCoordinator.kt
package com.luminet.android.tunnel

import java.net.InetAddress
import java.util.concurrent.ConcurrentHashMap

enum class AndroidMeshMode {
    DIRECT,
    DERP_RELAY
}

data class AndroidMeshPeer(
    val peerId: String,
    val publicKeyHex: String,
    val virtualIp: String,
    val allowedIps: List<String>,
    val endpoints: List<String>,
    val derpRegionId: Int,
    val lastHandshakeMs: Long,
    val isExitNode: Boolean
)

class MeshWireguardCoordinator(
    val localPeerId: String,
    val localVirtualIp: String
) {
    private val peers = ConcurrentHashMap<String, AndroidMeshPeer>()
    private val derpRelays = ConcurrentHashMap<Int, String>()

    fun registerDerpRelay(regionId: Int, host: String) {
        derpRelays[regionId] = host
    }

    fun registerPeer(peer: AndroidMeshPeer) {
        peers[peer.peerId] = peer
    }

    fun selectBestEndpoint(peerId: String, nowMs: Long): Pair<AndroidMeshMode, String>? {
        val peer = peers[peerId] ?: return null
        val isRecent = (nowMs - peer.lastHandshakeMs) < 180_000

        return if (isRecent && peer.endpoints.isNotEmpty()) {
            Pair(AndroidMeshMode.DIRECT, peer.endpoints.first())
        } else if (derpRelays.containsKey(peer.derpRegionId)) {
            Pair(AndroidMeshMode.DERP_RELAY, derpRelays[peer.derpRegionId]!!)
        } else if (peer.endpoints.isNotEmpty()) {
            Pair(AndroidMeshMode.DIRECT, peer.endpoints.first())
        } else {
            null
        }
    }

    fun lookupRoute(destIp: String): String? {
        // Exact virtual IP match
        peers.values.find { it.virtualIp == destIp }?.let { return it.peerId }

        // Subnet prefix match
        for (peer in peers.values) {
            for (cidr in peer.allowedIps) {
                if (matchesCidr(destIp, cidr)) {
                    return peer.peerId
                }
            }
        }
        return null
    }

    fun generatePeerConfig(peerId: String): String {
        val peer = peers[peerId] ?: throw IllegalArgumentException("Peer not found")
        val allowed = if (peer.allowedIps.isEmpty()) "${peer.virtualIp}/32" else peer.allowedIps.joinToString(", ")
        val endpointLine = if (peer.endpoints.isNotEmpty()) "Endpoint = ${peer.endpoints.first()}\nPersistentKeepalive = 25\n" else ""

        return """
            |[Peer]
            |PublicKey = ${peer.publicKeyHex}
            |AllowedIPs = $allowed
            |$endpointLine
        """.trimMargin().trim()
    }

    private fun matchesCidr(ipStr: String, cidrStr: String): Boolean {
        return try {
            val parts = cidrStr.split("/")
            if (parts.size != 2) return false
            val netAddr = InetAddress.getByName(parts[0]).address
            val targetAddr = InetAddress.getByName(ipStr).address
            val prefix = parts[1].toInt()
            if (prefix == 0) return true

            var matched = true
            var bits = prefix
            for (i in netAddr.indices) {
                if (bits >= 8) {
                    if (netAddr[i] != targetAddr[i]) { matched = false; break }
                    bits -= 8
                } else if (bits > 0) {
                    val mask = (0xFF shl (8 - bits)) and 0xFF
                    if ((netAddr[i].toInt() and mask) != (targetAddr[i].toInt() and mask)) { matched = false; break }
                    bits = 0
                } else break
            }
            matched
        } catch (e: Exception) {
            false
        }
    }
}
